src/basic/path-util.c

```diff
@@ -992,3 +993,52 @@ bool empty_or_root(const char *root) {
 
         return root[strspn(root, "/")] == 0;
 }
+
+int path_simplify_and_warn(
+                char *path,
+                unsigned flag,
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *lvalue) {
+
+        bool fatal, absolute;
+
+        assert((flag & (PATH_CHECK_ABSOLUTE | PATH_CHECK_RELATIVE)) != (PATH_CHECK_ABSOLUTE | PATH_CHECK_RELATIVE));
+
+        fatal = flag & PATH_CHECK_FATAL;
+
+        if (flag & (PATH_CHECK_ABSOLUTE | PATH_CHECK_RELATIVE)) {
+                absolute = path_is_absolute(path);
+
+                if (!absolute && (flag & PATH_CHECK_ABSOLUTE)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0,
+                                   "%s= path is not an absolute%s: %s",
+                                   fatal ? "" : ", ignoring", lvalue, path);
+                        return -EINVAL;
+                }
+
+                if (absolute && (flag & PATH_CHECK_RELATIVE)) {
+                        log_syntax(unit, LOG_ERR, filename, line, 0,
+                                   "%s= path is an absolute%s: %s",
+                                   fatal ? "" : ", ignoring", lvalue, path);
+                        return -EINVAL;
+                }
+        }
+
+        path_kill_slashes_and_dot(path, true);
+
+        if (!path_is_normalized(path)) {
+                log_syntax(unit, LOG_ERR, filename, line, 0,
+                           "%s= path is not normalized%s: %s",
+                           fatal ? "" : ", ignoring", lvalue, path);
+                return -EINVAL;
+        }
+
+        if (!utf8_is_valid(path)) {
```

Comment:
So this check must be first, because otherwise we can print non-utf8 in the messages above.

Reply:
OK. it is moved to the beginning of the function.